Add Beat input plugin

nferch opened the pull request:
### Required for all PRs:

- [X] Signed [CLA](https://influxdata.com/community/cla/).
- [X] Associated README.md updated.
- [X] Has appropriate unit tests.

As discussed in https://github.com/influxdata/telegraf/pull/4911, combine @dmitryilyin 's filebeat and kafkabeat plugins.

nferch commented:
@danielnelson any thoughts or feedback?

danielnelson commented:
I'll have to get back to you on a full review.  Quick look though overall the code is simple and straightforward, though I'd rather avoid FlattenJSON and just pick the values out manually.  I'm also unsure of the utility of collecting the system metrics via this method.  Since all the stats are in one response I also suggest we remove the options about what sections to collect.

nferch commented:
Filebeat exposes two different URIs, `/` and `/stats`. `/stats` is the only one I really care about, inclusion of the other metrics was based on @esity's comment in https://github.com/influxdata/telegraf/pull/4911 to make a generalized Beats plugin.

Can't speak to much to the implementation as this PR is largely @dmitryilyin's code, but can change from FlattenJSON at the cost of tying the implementation to the API response du jour.

nferch commented:
I have been using this PR in limited production and it satisfies my use case, detecting queues/backpressure in Filebeat using the `beat_filebeat.events_active` measurement. @dmitryilyin what was your use case? Do you have a Kafka instance to test against?

axelrtgs commented:
I have been eagerly watching this and waiting for the feature to come to telegraf. I wanted to leave a comment because i believe the current implementation would also work for my use case

Esity commented:
@danielnelson Just wondering where we are at with this one. Looks pretty good to me

nferch commented:
@danielnelson Happy New Year! :) Any further thoughts or feedback?

axelrtgs commented:
@danielnelson Hope all is well, Been eagerly waiting for this to be implemented. Was wondering when you would have a minute to look at it or if there was anything those of us waiting could do to help speed the process along?

axelrtgs commented:
@danielnelson @sjwang90 I know you guys have a lot on your plate but this seems to be ready since November. Every few months it keeps getting bumped to the next version  Is there anything I can do to get this looked at and make it easier for you to approve and maybe actually get it into the 1.15 release?

alsargent commented:
Glad we have this on our radar. Just looking at Elastic Beats plugins (aka "community beats"), there are a number of technologies that Telegraf doesn't yet cover: https://www.elastic.co/guide/en/beats/libbeat/current/community-beats.html

nferch commented:
@sjwang90 re https://github.com/influxdata/telegraf/pull/4911#issuecomment-730862173, any update or end in sight on this PR? It's been open for over a year with several missed milestone inclusions.

Happy to take a look at the code and ensure it's up to date with the current master branch if there's some indication that'd be time worth spending.

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,278 @@
+const sampleConfig = `
+  ## An URL from which to read Beat-formatted JSON
+  ## Default is "http://127.0.0.1:5066".
+  url = "http://127.0.0.1:5066"
```
Would it make sense to collect from multiple hosts? Then this should be a list...

nferch replied:
Think this would be of limited use, I imagine the most common deployments would be a local Telegraf agent and Beat agent per host. So I'd rather address this in a future PR if there's a need that isn't satisfied by a separate configuration per host.

srebhan replied:
Fair enough, please make as resolved.

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,278 @@
+  url = "http://127.0.0.1:5066"
+
+  ## Enable collection of the Beat stats
+  collect_beat_stats = true
+
+  ## Enable the collection if Libbeat stats
+  collect_libbeat_stats = true
+
+  ## Enable the collection of OS level stats
+  collect_system_stats = false
+
+  ## Enable the collection of Filebeat stats
+  collect_filebeat_stats = true
```
Please choose a good default e.g. all is collected and keep the number of uncommented sample config options low. The idea is that one can start by just passing the host and might later customize with the options.

nferch replied:
Will fix

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,278 @@
+  # host_header = "logstash.example.com"
+
+  ## Timeout for HTTP requests
+  timeout = "5s"
```
Same. Please comment this option with the (good) default to not confuse new users.

nferch replied:
Will fix

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,278 @@
+	client *http.Client
+}
+
+func NewBeat() *Beat {
+	return &Beat{
+		URL:                  "http://127.0.0.1:5066",
+		CollectBeatStats:     true,
+		CollectLibbeatStats:  true,
+		CollectSystemStats:   true,
+		CollectFilebeatStats: true,
+		Method:               "GET",
+		Headers:              make(map[string]string),
+		HostHeader:           "",
+		Timeout:              internal.Duration{Duration: time.Second * 5},
+	}
+}
```
Make this an `Init()` function and later (in `init()`) pass an empty config.

nferch replied:
Will fix

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,278 @@
+		return nil, err
+	}
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsConfig,
+		},
+		Timeout: beat.Timeout.Duration,
+	}
```
How about e.g. proxy settings?

nferch replied:
Couldn't find many other plugins that support proxies, was there something about this one in particular that you thought necessitated it?

srebhan replied:
The http plugin supports proxies for examples. The reason is that I had quite some software _not_ accepting proxies in the past and that might be a nightmare in enterprise environments. I don't think adding those is super hard either.

nferch replied:
From what I can tell, the http plugin is the only other plugin that supports proxies. I could see adding it if there's enough prevailing need or consistency with the other plugins that support it. Seems like something that could be easily added if necessary.

srebhan replied:
I take your word on this! ;-) If you really don't want to do it now, mark this as resolved...

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,278 @@
+// gatherJsonData query the data source and parse the response JSON
+func (beat *Beat) gatherJsonData(url string, value interface{}) error {
+
+	var method string
+	if beat.Method != "" {
+		method = beat.Method
+	} else {
+		method = "GET"
+	}
```
This default should be set in `Init()`.

nferch replied:
will fix

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,278 @@
+		return err
+	}
+
+	if (beat.Username != "") || (beat.Password != "") {
```
Does it make sense to have a password with an empty username? Does this work at all?

nferch replied:
will fix

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,278 @@
+	if (beat.Username != "") || (beat.Password != "") {
+		request.SetBasicAuth(beat.Username, beat.Password)
+	}
+	for header, value := range beat.Headers {
```
The local variable `value` might shadow the function argument `value`. To be honest I'm also not sure that I like the passing the result `value` by reference instead of constructing it here and return it. Is there any reason for doing this?

nferch replied:
not sure if I understand the latter part of your comment, this bit is just applying the headers configuration. will rename the variables from the range to avoid them clashing with the value function parameter, LMK if there's other concerns.

srebhan replied:
What I mean is instead of passing `value` to `gatherJsonData()` you could do `value := &BeatInfo{}` *in* that function and then return `value` to the caller.

nferch replied:
ahh yeah, that is a bit weird. I'm not the original author so I don't have the full context as far as history or intentions, but I'll fix this one.

nferch replied:
actually, I understand now. :) It's passed a `BeatInfo` and later a `BeatStats`, so it needs to accept an `interface{}`

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,278 @@
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/internal/tls"
```
This is 
```suggestion
	"github.com/influxdata/telegraf/plugins/common/tls"
```
now I think.

nferch replied:
will fix

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,278 @@
+
+	defer response.Body.Close()
+
+	err = json.NewDecoder(response.Body).Decode(value)
+	if err != nil {
+		return err
+	}
+
+	return nil
```
You might directly return `err` here.

nferch replied:
will fix

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,278 @@
+}
+
+func (beat *Beat) gatherInfoTags(url string) (map[string]string, error) {
+	beatInfo := &BeatInfo{}
+
+	err := beat.gatherJsonData(url, beatInfo)
+	if err != nil {
+		return nil, err
+	}
+
```
Couldn't you move this out of this function so that we have the same structure below?

nferch replied:
will fix

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,278 @@
+	if err != nil {
+		return err
+	}
+
```
Maybe call `beat.gatherJsonData(infoUrl.String(),...)` here and let `beat.gatherInfoTags(...)` only scrape the tags from the return. This then has all the "shoot a query and parse the return" functions on the same level.

nferch replied:
completely flattened `gatherInfoTags` since it doesn't actually do that much. think this reads much clearer, but open to other ideas.

srebhan replied:
Please mark as resolved.

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,278 @@
+
+	if beat.CollectBeatStats {
+		flattenerBeat := jsonparser.JSONFlattener{}
+		err := flattenerBeat.FlattenJSON("", beatStats.Beat)
```
According to the comment, `FlattenJSON()` does *NOT* handle booleans and strings. Do you maybe want `FullFlattenJSON()` here?

nferch replied:
will fix

srebhan replied:
Still open...

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,278 @@
+	}
+
+	if beat.CollectFilebeatStats {
+		flattenerBeat := jsonparser.JSONFlattener{}
```
Maybe you can have one instance of the `flattenerBeat` and clear the fields between the calls...

nferch replied:
only 4 instances so I'll just leave it as-is, but open to changing it if there's considerable efficiency to be gained...

srebhan replied:
Hmm. I don't see the reason not to declare `flattener := jsonparser.JSONFlattener{}` and use it in the `if`s. The flattener is even stateless, so you save the construction lines... :-)

nferch replied:
I'm not too familiar with the API, but that doesn't seem to be the case. Simply making the additional calls use the same instance without clearing it (is there a convenient way to do that?) causes the tests to fail.

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,278 @@
+		if err != nil {
+			return err
+		}
+		accumulator.AddFields("beat_filebeat", flattenerBeat.Fields, tags)
```
Your metric name `beat_filebeat` does not follow the convention (default: plugin name). This is especially bad if people use `name_override` to customize the output and suddenly don't get the three distinct streams but an undistinguishible mixture.
How about adding the "information type" (`filebeat` in this case) as a tag (e.g. `source="filebeat"`)?

nferch replied:
my thinking here was that it made more sense to have separate measurements as there's not much overlap in the responses, it's the interface that is consistent between the endpoints.

don't really have a strong opinion on it though.

srebhan replied:
Go ahead as-is for now. I have to discuss this more with Influx-people... Many plugins do this and if you don't be careful you easily screw up the metrics when using `name_override` as it doesn't care about the suffix...
Mark resolved please.

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,278 @@
+		accumulator.AddFields("beat_filebeat", flattenerBeat.Fields, tags)
+	}
+
+	if beat.CollectLibbeatStats {
+		flattenerLibbeat := jsonparser.JSONFlattener{}
+		err := flattenerLibbeat.FlattenJSON("", beatStats.Libbeat)
+		if err != nil {
+			return err
+		}
+		accumulator.AddFields("beat_libbeat", flattenerLibbeat.Fields, tags)
+	}
+
+	if beat.CollectSystemStats {
+		flattenerSystem := jsonparser.JSONFlattener{}
+		err := flattenerSystem.FlattenJSON("", beatStats.System)
+		if err != nil {
+			return err
+		}
+		accumulator.AddFields("beat_system", flattenerSystem.Fields, tags)
+	}
```
Same comments as for the `beat.CollectBeatStats` section.

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,278 @@
+}
+
+func (beat *Beat) Gather(accumulator telegraf.Accumulator) error {
+	if beat.client == nil {
+		client, err := beat.createHttpClient()
+
+		if err != nil {
+			return err
+		}
+		beat.client = client
+	}
```
Could be done in `Init()`.

nferch replied:
will fix

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,278 @@
+
+func init() {
+	inputs.Add("beat", func() telegraf.Input {
+		return NewBeat()
```
```suggestion
		&Beat{}
```
and setting the defaults in an `Init()` function.

nferch replied:
will fix

srebhan commented on plugins/inputs/beat/beat_test.go:
```diff
@@ -0,0 +1,225 @@
+
+var beatTest = NewBeat()
+
+var (
+	beat6StatsAccumulator testutil.Accumulator
+)
```
Please move this to the individual tests.

nferch replied:
will fix

srebhan commented on plugins/inputs/beat/beat_test.go:
```diff
@@ -0,0 +1,225 @@
+	"github.com/stretchr/testify/assert"
+)
+
+var beatTest = NewBeat()
```
Please move this to the individual tests.

nferch replied:
will fix

srebhan commented on plugins/inputs/beat/beat_test.go:
```diff
@@ -0,0 +1,225 @@
+	fakeServer.Start()
+	defer fakeServer.Close()
+
+	if beatTest.client == nil {
+		client, err := beatTest.createHttpClient()
+
+		if err != nil {
+			test.Logf("Can't createHttpClient")
+		}
+		beatTest.client = client
+	}
```
Why is this necessary? The beat-plugin should connect on its own no?

nferch replied:
will fix

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,278 @@
+	for header, value := range beat.Headers {
+		request.Header.Add(header, value)
+	}
+	if beat.HostHeader != "" {
+		request.Host = beat.HostHeader
+	}
```
Can also be moved to `Init()`.

nferch replied:
I don't understand, this is setting `Host` on the HTTP request object, which is instantiated per `gatherJsonData` call.

srebhan replied:
You are absolutely right. Sorry missed this one. Mark as resolved.

srebhan commented on plugins/inputs/beat/beat_test.go:
```diff
@@ -0,0 +1,225 @@
+	//beat6StatsAccumulator.SetDebug(true)
+	fakeServer := httptest.NewUnstartedServer(
+		http.HandlerFunc(
+			func(writer http.ResponseWriter, request *http.Request) {
+				writer.Header().Set("Content-Type", "application/json")
+				if request.URL.String() == suffixInfo {
+					fmt.Fprintf(writer, "%s", string(beat6Info))
+				} else if request.URL.String() == suffixStats {
+					fmt.Fprintf(writer, "%s", string(beat6Stats))
+				} else {
+					test.Logf("Unkown URL: " + request.URL.String())
+				}
+			},
```
The test-data should move to a JSON file you load and dump over HTTP. In the current schema you rely on the compiler to detect that the consts can be compiled out which is... not so nice. :-)

Look at other current test schemes (e.g. into the starlark processor or the `inputs/rabbitmq` tests) to get an idea how this could look like.

nferch replied:
will fix

srebhan commented on plugins/inputs/beat/sample_version6.go:
```diff
@@ -0,0 +1,151 @@
+package beat
+
+const beat6Info = `
```
Should be a JSON file. Really. :-)

nferch replied:
will fix

srebhan commented on plugins/inputs/beat/sample_version6.go:
```diff
@@ -0,0 +1,151 @@
+}
+`
+
+const beat6Stats = `
```
Same here.

nferch replied:
will fix

srebhan reviewed (changes requested):
Sorry for this taking soooo long!

I have some comments and like you to consider introducing an `Init()` function to set the defaults and one-time OPs there. This could simplify the code.
Furthermore, please rework the tests. Check e.g. `rabbitmq` on how to keep and use the testdata as JSON files.
Beside this there are some more comments in the code...

Let me know if you have questions.

srebhan commented:
@nferch any news on this?

nferch commented:
@srebhan from my cursory review the comments make sense, should have some time this weekend to work on it.

nferch commented:
@srebhan ok, I think I addressed all of your comments. LMK if there's anything I missed.

Would welcome any other patches or real-world testing, as I'm not the original author of the code and only have Filebeat running to test it against.

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -78,10 +78,10 @@ type BeatStats struct {
-	CollectLibbeatStats  bool `toml:"collect_libbeat_stats"`
-	CollectSystemStats   bool `toml:"collect_system_stats"`
-	CollectFilebeatStats bool `toml:"collect_filebeat_stats"`
+	CollectBeatStats     *bool `toml:"collect_beat_stats"`
+	CollectLibbeatStats  *bool `toml:"collect_libbeat_stats"`
+	CollectSystemStats   *bool `toml:"collect_system_stats"`
+	CollectFilebeatStats *bool `toml:"collect_filebeat_stats"`
```
Why do you need those to be pointers?

nferch replied:
Since we are setting defaults in `Init()` after the config has been read, it would not be possible to discern a false setting from an unset setting. Open to other ways to achieve this, this is how I've seen it done in other projects.

ssoroka replied:
set the defaults in init(), and then you don't have to check them at all, and the fields don't have to be pointers.

nferch replied:
@ssoroka This is how it was prior to the comments on https://github.com/influxdata/telegraf/pull/6653#discussion_r530877623

ssoroka replied:
Sorry about that. the correct way to set defaults is when you create the plugin, (around the lowercase `init()` call), By the time `Init()` is called, the config has been loaded over top of the defaults, and has the correct settings. Trying to set them in `Init()` leads to some weirdness, as you've experienced. I'd also like to go ahead with the change to change the bools to a `[]string` slice, because once we merge it the config is locked and we can't break backwards compatibility.

srebhan replied:
@ssoroka this is my fault. I thought it is cleaner to initialize the defaults in `Init()` but completely missed that you cannot know if the boolean was specified as false by the user or it is false due to not being initialized. Again, my fault. Sorry @nferch for the confusion I caused!

nferch replied:
Ok, I've reverted the flow to set defaults in `init()`.

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -94,18 +94,49 @@ type Beat struct {
+	defaultMethod = "GET"
+)
+
+var defaultCollectBeatStats = true
+var defaultCollectLibbeatStats = true
+var defaultCollectSystemStats = false
+var defaultCollectFilebeatStats = true
+var defaultTimeout = internal.Duration{Duration: time.Second * 5}
```
Please don't declare them here. Either you declare those in `Init()` or you directly set the defaults there.

nferch replied:
Was necessary to make them vars in order to take the address of them. Thought it was cleaner to separate the default values from the code of setting the defaults.

srebhan replied:
See my comment to change to a collection list in the form `collect = ["beat", "libbeat",...]`. You can use `internal/choice` to check the options an see what is in the user-setting.

srebhan commented on plugins/inputs/beat/README.md:
```diff
@@ -0,0 +1,149 @@
+  url = "http://127.0.0.1:5066"
+
+  ## Enable collection of the generic Beat stats
+  collect_beat_stats = true
+
+  ## Enable the collection if Libbeat stats
+  collect_libbeat_stats = true
+
+  ## Enable the collection of OS level stats
+  collect_system_stats = false
+
+  ## Enable the collection of Filebeat stats
+  collect_filebeat_stats = true
+
+  ## HTTP method
+  # method = "GET"
+
+  ## Optional HTTP headers
+  # headers = {"X-Special-Header" = "Special-Value"}
+
+  ## Override HTTP "Host" header
+  # host_header = "logstash.example.com"
+
+  ## Timeout for HTTP requests
+  timeout = "5s"
```
You might also want to comment the default to match `sampleConfig`.

nferch replied:
good catch, will fix.

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,268 @@
+var defaultTimeout = internal.Duration{Duration: time.Second * 5}
+
+func (beat *Beat) Init() error {
+	var err error
+	beat.client, err = beat.createHTTPClient()
```
```suggestion
	beat.client, err := beat.createHTTPClient()
```

nferch replied:
unless I'm mistaken, it's necessary to declare separately since `:=` will only work for new variables or variables defined in the current block. The suggested change does not compile.

srebhan replied:
Hmmm what is the error message? It should work I think...

nferch replied:
`plugins/inputs/beat/beat.go:109:6: non-name beat.client on left side of :=`

srebhan replied:
Sorry, missed the structure part... Should sleep more. :-( Keep as is and mark my this one as resolved.

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,268 @@
+
+	defer response.Body.Close()
+
+	err = json.NewDecoder(response.Body).Decode(value)
+	return err
```
You can also do
```suggestion
	return json.NewDecoder(response.Body).Decode(value)
```
Both ok for me.

srebhan commented on plugins/inputs/beat/beat.go:
```diff
@@ -0,0 +1,268 @@
+  url = "http://127.0.0.1:5066"
+
+  ## Enable collection of the Beat stats
+  # collect_beat_stats = true
+
+  ## Enable the collection if Libbeat stats
+  # collect_libbeat_stats = true
+
+  ## Enable the collection of OS level stats
+  # collect_system_stats = false
+
+  ## Enable the collection of Filebeat stats
+  # collect_filebeat_stats = true
```
Concerning our "default value" discussion below, it might be easier to allow the user to specify a list, i.e.
```
collect =["beat", "libbeat", "filebeat"]
```
This way you know if an option was selected (it's in the list) or not. Maybe you can define an empty list to "collect all"...

srebhan replied:
This is also more extensible in the future if you want to e.g. increase granularity...

nferch replied:
I don't have a strong opinion on this, but don't feel much motivation to change it either. The design decision to use booleans predated my commits which sought to integrate the plugins in the simplest way possible. The use of booleans to toggle functionality seemed similar to other plugins so it didn't seem unusual to use them in this way. I don't see how it would make a difference to the end user either, but I obviously have a limited perspective on this.

srebhan replied:
Yeah it's similar to the plugins where we currently have a lot of trouble extending them without breaking configs... Many plugins therefore move to `include`, `exclude` lists to allow sophisticated filtering of metrics. This is nice for the end-user for cases where you want all but one.

nferch replied:
So is this just a "nice to have" or a blocker for moving forward? I can see how it'd be useful for some more complicated use cases or extensibility, but I don't have much reason to believe that how it is currently implemented doesn't solve the problems of the present. And I'm reluctant to spend much time solving the problems of the future, when they can be easily addressed via a PR once they materialize.

I'd assume that the prospective users of this plugin would prefer having a potentially less-perfect solution sooner than a more future proof solution later, but I defer to anyone else who is following this thread to chime in with their use cases and preferences, as all I have to go by on is my personal one.

axelrtgs replied:
I would vote for nice to have vs blocker. i know i would love to see this out so i can monitor filebeats status even if its not perfect. Although if its not too much work it may be reasonable to do it now since i know fixing things later rarely happens and although i would love to contribute i have 0 knowledge of GO to even pick this up myself in a second PR.

Perhaps merging this functional PR as is with the non perfect config but then opening an issue or second PR to enhance the config may be a compromise between both options.

srebhan replied:
The problem is that we **cannot** fix this *later* as we need to keep backward compatible with the config. Instead of arguing over this again I implemented what is necessary in less time than the discussion took so far...
Please see my comment below with the attached patch (`git apply ...`). Feel free to use it!

nferch replied:
Thanks, applied.

srebhan commented:
Here is my [implementation](https://github.com/influxdata/telegraf/files/5705290/beat-plugin-include-list.patch.txt)  of converting the booleans to a list of includes... Feel free to use it.

nferch commented:
@srebhan @ssoroka I believe I've addressed all the comments and feedback. LMK if there's anything else, otherwise I'll squash and rebase so it's ready for merge.

srebhan reviewed (approved):
Looks good to me now! Also thx for further cleanup!

srebhan commented:
@ssoroka can you maybe add a second opinion as I provided the include-list-part before flagging this ready-to-merge!?

ssoroka reviewed (approved):
lgtm :)

ssoroka commented:
thanks for the epic effort here. much appreciated
